rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1210 @@
+  This allows the formatter to expand or compact multi-line expressions, while still allowing grouping of code.
+
+  For example, formatting this code:
+  ```nix
+  [
+    0 10
+
+    (
+      20 + 1
+    )
+
+
+    30
+  ]
+  ```
+
+  turns into this:
+  ```nix
+  [
+    0  # Line break added
+    10
+
+    (20 + 1) # Line breaks removed
+             # Consecutive empty lines turned into a single empty line
+    30
+  ]
+  ```
+
+- Expressions of the same kind that can be treated as a sequence of expressions on the same level should be treated as such, even though they are technically parsed as a nested tree.
+  - This applies to else-if chains, functions with multiple arguments, some operators, etc.
+  - Example:
+    ```nix
+    # This is treated as a sequence of if-then-elsa's chains
+    if cond1 then
+      foo
+    else if cond2 then
+      bar
+    else
+      baz
+    ```
+
+- Indentation should reflect the expression structure.
+  Example:
+  ```nix
+  # Bad, the indentation misleads the user
+  {
+    foo = {
+    bar = if
+    baz == null then 10
+      else 20
+    ;
+  }; }
+
+  # Good
+  {
+    foo = {
+      bar =
+        if baz == null then
+          10
+        else
+          20;
+    };
+  }
+  ```
+
+### Editor Config
+
+This [editor config](https://editorconfig.org/) specifies the basic details about Nix files:
+
+```toml
+end_of_line = lf
+insert_final_newline = true
+trim_trailing_whitespace = true
+charset = utf-8
+indent_style = space
+```
+
+### Single-line common ancestor expression rule
+
+For any two (sub-)expressions that are fully on a common single line, their smallest common ancestor expression must also be on the same line.
+
+**Example**
+
+```nix
+# Bad, expressions cond and bar are fully on the same line,
+# but their smallest common ancestor expression is the entire if-then-else, which spans multiple lines
+if cond then foo
+else bar
+
+# Okay, cond, foo and bar have the if-then-else as a common ancestor expression,
+# which is also fully on the same line
+if cond then foo else bar
+
+# Bad, due to function application precedence, the smallest common ancestor expression
+# of foo and bar is `foo || bar baz`, which spans two lines
+foo || bar
+  baz
+```
+
+**Rationale**
+
+This rule has turned out to be very practical at catching code that could be potentially hard to understand or edit.
+
+### Line length
+
+- There should be a configurable _soft_ line length limit, limiting the number of characters on one line without counting the leading indentation.
+  The default should be 100 characters.
+- There may also be a configurable _hard_ line length limit, which includes the leading indentation.
+- String-like values such as strings, paths, comments, urls, etc. may go over the hard line length limit.
+
+### Indentation
+
+- Two spaces must be used for each indentation level.
```

Comment:
@toastal I definitely have the same issue. In my personal Nix files repo I set a wider indentation (with tabs) because it was too difficult to keep track of indentation otherwise. Often when reading Nixpkgs code I have to reformat the indentation or add comments that note different ends of indentation blocks to be able to understand what I'm looking at. It's definitely not very accessible to me, at least.

Reply:
I'm curious if structure-guide lines help you at all.

Reply:
Surprisingly not really. It can help a little in some cases but it also adds more visual noise.

Reply:
@toastal if you use vscode or vscodium, you can use this extension to fix it on your end: https://marketplace.visualstudio.com/items?itemName=kylepaulsen.stretchy-spaces

i have not found anything similar for vim yet tho

Reply:
While these are good arguments, 2 spaces of indentation is by far the most popular style right now, and we consider changing this out of scope for this RFC. Furthermore, the next sentence in the RFC states:

> This may be revisited should Nix get proper support for [using tabs for indentation](https://github.com/NixOS/nix/issues/7834) in the future.

Reply:
Currently it's de-facto standard to use 2 spaces of indentation, that's what everybody has specified in their editor configs and has been using for the past 20 years. Changing this would be a substantial change on its own and deserve its own RFC. I even think it would be deceiving to now include this in the RFC after it has gone all this time without even mentioning this. So for anybody that's passionate about this, please open a separate RFC!

Generally, to make an RFC succeed, it should be focused and as minimal as possible. So really, we shouldn't even mention how indentation is done at all, it's really not very relevant for this RFC.